File: 70. Climbing Stairs/70.js
// 70. Climbing Stairs solution in JavaScript
// https://leetcode.com/problems/climbing-stairs/description/

/**
 * @param {number} n
 * @return {number}
 */
var climbStairs = function(n) {
    let memo = new Array(n + 1).fill(0);

    return climb_options(0, n, memo);
};

function climb_options(i, n, memo) {
    if (i > n) return 0;

    if (i == n) return 1;

    if (memo[i] > 0) {
        return memo[i];
    }

    memo[i] = climb_options(i + 1, n, memo) + climb_options(i + 2, n, memo);

    return memo[i];
}

File: 70. Climbing Stairs/70.go
// 70. Climbing Stairs solution in Golang
// https://leetcode.com/problems/climbing-stairs/description/

func climbStairs(n int) int {
    if n == 1 {
        return 1;
    }

    var dp = make([]int, n + 1)

    dp[1] = 1
    dp[2] = 2

    for i := 3; i <= n; i++ {
        dp[i] = dp[i - 1] + dp[i - 2];
    }

    return dp[n];
}